### PR TITLE
[2.13.x] DDF-{4365, 4370, 4372, 4374} Minor bugfixes/improvements

### PR DESCRIPTION
**Pull request by SmithJosh**
#### What does this PR do?
* Fixes PAOS Interceptor to support backend HTTP DELETE requests to secured REST. This required changing the interceptor to compute the request content length correctly.
* Cleans up old source polling events when changes are made in the Sources UI. Previously, old events weren't removed and while using the Sources UI outdated events would spam the backend.
* Resolves an AccessControlException in the subscriptions:list command.
* Improves the output of the subscriptions:list command so an admin can tell whether it's an enterprise subscription or a source-specific subscription.

#### Who is reviewing it? 
@bakejeyner @blen-desta @Kjames5269 @brjeter 
#### Ask 2 committers to review/merge the PR and tag them here.

@coyotesqrl
#### How should this be tested?
TEST 1: Sources UI
1. Create a new source in the sources UI, and open the browser dev tools to view the network tab.
2. Select and close the source edit modal a few times, and disable/re-enable the source a few times.
3. Verify there is still just a single request to the backend every 30 seconds.
4. Disable the source. Verify that the requests to the backend stop

TEST 2: subscriptions:list AccessControlException
1. Create a subscription pointing to a site that DDF does not trust, e.g. example.com
2. Run the subscriptions:list command and verify that there is no AccessControlException

#### What are the relevant tickets?
[DDF-4365](https://codice.atlassian.net/browse/DDF-4365)
[DDF-4370](https://codice.atlassian.net/browse/DDF-4370)
[DDF-4372](https://codice.atlassian.net/browse/DDF-4372)
[DDF-4374](https://codice.atlassian.net/browse/DDF-4374)
#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

**Comment by SmithJosh** [comment]
I had a few other tickets that required minor changes to DDF, so instead of creating PRs for all of them, I just included them here.

**Comment by coyotesqrl** [comment]
@SmithJosh Knowing how muscle memory works, whoever commits this will likely `Squash and merge` even if they don't mean to do so. That'll collapse the separate logical commits in a way that will lose all context and sense. I know it's a bit more effort, but these should be broken up into separate PRs.

**Comment by coyotesqrl** [comment]
After speaking with @SmithJosh offline, we've decided it's best to keep these together and pay special attention NOT to squash at merge time.

File: ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js [hunk 1]
```diff
@@ -33,18 +33,28 @@ define([
         Source.Model = Backbone.Model.extend({
             configUrl: "../jolokia/exec/org.codice.ddf.ui.admin.api.ConfigurationAdmin:service=ui",
             idAttribute: 'name',
-            defaults: function() {
-              return {
-                currentUrl:     undefined,
-                isLoopbackUrl:   undefined,
-              };
+            defaults: {
```
Comment (coyotesqrl) [inline ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js]: ❓ Not `function()`?

File: catalog/core/catalog-core-eventcommands/src/main/java/org/codice/ddf/catalog/pubsub/command/ListCommand.java [hunk 2]
```diff
@@ -73,46 +80,95 @@
...
-    Subscription subscription = bundleContext.getService(entry.getValue());
-    String rowColor = "";
 
-    if (subscription != null
-        && subscription.getDeliveryMethod() instanceof Pingable
-        && !((Pingable) subscription.getDeliveryMethod()).ping()) {
-      rowColor = RED_CONSOLE_COLOR;
-    }
+    int[] columnWidth = new int[4];
+    columnWidth[0] = entries.stream().mapToInt(s -> s.id.length()).max().orElse(-1);
```
Comment (coyotesqrl) [inline catalog/core/catalog-core-eventcommands/src/main/java/org/codice/ddf/catalog/pubsub/command/ListCommand.java]: ✏️ A loop would be more efficient here than these four separate reduce operations.

File: ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js (hunk 1 above)
Comment (SmithJosh) [reply ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js]: None of the attribute values are objects, so the `function()` syntax isn't necessary. See https://stackoverflow.com/q/17318048

File: catalog/core/catalog-core-eventcommands/src/main/java/org/codice/ddf/catalog/pubsub/command/ListCommand.java [hunk 3]
```diff
@@ -73,46 +80,95 @@
...
+            (subscription.getSourceIds() == null)
+                ? "[]"
+                : Arrays.toString(subscription.getSourceIds().toArray());
+      }
+    }
+
+    public boolean ping() {
+      return (subscription != null)
+          && (subscription.getDeliveryMethod() instanceof Pingable)
+          && ((Pingable) subscription.getDeliveryMethod()).ping();
```
Comment (brjeter) [inline catalog/core/catalog-core-eventcommands/src/main/java/org/codice/ddf/catalog/pubsub/command/ListCommand.java]: ❓ Technically this condition is different than what it was before. For instance, this will return false (and color the row red) if the subscription isn't an instance of `Pingable` whereas before it would only color the row red if the entry **_was_** an instance of `Pingable` **and** the ping failed. Was this change intentional?

Comment (SmithJosh) [reply catalog/core/catalog-core-eventcommands/src/main/java/org/codice/ddf/catalog/pubsub/command/ListCommand.java]: That was not intentional. I'll change it to be consistent with the existing behavior

File: ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js [hunk 4]
```diff
@@ -311,7 +326,7 @@ define([
             parseServiceModel: function () {
                 var resModel = this;
                 var collection = resModel.get('collection');
-                collection.reset();
+                collection.removeAllSources();
```
Comment (andrewkfiedler) [inline ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js]: What's the motivation to switch here?

File: ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js [hunk 5]
```diff
@@ -57,29 +67,29 @@ define([
...
                 } else if (configuration === this.get("currentConfiguration")) {
                     this.stopListening(configuration);
+                    this.stopPolling();
                     this.set("currentConfiguration", undefined);
                 }
             },
             setCurrentConfiguration: function (configuration) {
                 var sync = 'sync';
                 var statusUpdate = 'status:update-';
-                this.set({
```
Comment (andrewkfiedler) [inline ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js]: Were we just setting the configuration twice? 😄

File: ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js [hunk 6]
```diff
@@ -57,29 +67,29 @@ define([
...
                 var pid = configuration.id;
-                var statusModel = new Status.Model(pid);
-                statusModel.on(sync, function () {
-                    wreqr.vent.trigger(statusUpdate + pid, statusModel);
-                });
 
-                var options = {
-                    delay: 30000
-                };
+                var statusModel = this.get('statusModel');
```
Comment (andrewkfiedler) [inline ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js]: Is this updating it to avoid making multiple status models?

File: ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js [hunk 7]
```diff
@@ -33,18 +33,28 @@ define([
...
-              };
+            defaults: {
+                currentUrl: undefined,
+                isLoopbackUrl: undefined,
             },
             initialize: function () {
                 this.set('currentConfiguration', undefined);
                 this.set('disabledConfigurations', new Source.ConfigurationList());
                 this.listenTo(this, 'change:currentConfiguration', this.updateCurrentBinding);
+                this.listenTo(this, 'remove', this.onRemove);
```
Comment (andrewkfiedler) [inline ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js]: I'm not as familiar with this code.  What is the remove event here?  This looks like a model to me, not a collection, so I'm a slightly confused.

File: ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Status.js [hunk 8]
```diff
@@ -13,19 +13,24 @@
...
-        url: "../jolokia/exec/org.codice.ddf.catalog.admin.poller.AdminPollerServiceBean:service=admin-source-poller-service/sourceStatus/",
-            model: Status.Model,
+        url: undefined,
+        model: Status.Model,
 
-            initialize: function(pid) {
-                this.url += pid;
-            }
-        });
+        initialize: function(pid) {
```
Comment (andrewkfiedler) [inline ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Status.js]: I think this model could be updated to use http://backbonejs.org/#Model-urlRoot

That in combination with setting id to pid should set the url correctly.

File: ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js (hunk 4 above)
Comment (SmithJosh) [reply ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js]: `collection.reset()` doesn't trigger remove events on the source models, and I changed the source model to clean up polling events when removed.

File: ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js (hunk 5 above)
Comment (SmithJosh) [reply ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js]: Yep 🤦‍♂️

File: ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js (hunk 6 above)
Comment (SmithJosh) [reply ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js]: Exactly. It would create a new model every time `setCurrentConfiguration` was called. And the old model would still be happily polling away 😄 I changed the source only uses a single model, and in turn, a single poller that can be cleaned up when the source is removed

File: ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js (hunk 7 above)
Comment (SmithJosh) [reply ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Source.js]: The 'remove` event gets triggered when calling `collection.remove()`. See http://backbonejs.org/#Collection-remove

File: ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Status.js (hunk 8 above)
Comment (SmithJosh) [reply ui/packages/catalog-admin-module-sources/src/main/webapp/js/model/Status.js]: 👍 Didn't know about that. I can change it

**Comment by SmithJosh** [comment]
Fixed merge conflicts and addressed comments. This is ready for re-review

File: platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/paos/PaosInInterceptor.java [hunk 9]
```diff
@@ -341,14 +339,15 @@ HttpUnsuccessfulResponseHandler getHttpUnsuccessfulResponseHandler(Message messa
...
-              new InputStreamContent(
-                  (String) message.get(Message.CONTENT_TYPE),
-                  new ByteArrayInputStream(byteArrayOutputStream.toByteArray()));
+          try (TemporaryFileBackedOutputStream tfbos = new TemporaryFileBackedOutputStream()) {
+            message.setContent(OutputStream.class, tfbos);
+            BodyWriter bodyWriter = new BodyWriter();
+            bodyWriter.handleMessage(message);
+            ByteSource byteSource = tfbos.asByteSource();
+            content =
+                new InputStreamContent(method, byteSource.openStream())
```
Comment (brjeter) [inline platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/paos/PaosInInterceptor.java]: ❓ Should we be passing in the HTTP Method as the "type" parameter instead of `(String) message.get(Message.CONTENT_TYPE)`

Comment (SmithJosh) [reply platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/paos/PaosInInterceptor.java]: Yes!

**Comment by blen-desta** [comment]
✅ Hero's successfully. 
- Verified that source requests are sent to the backend every 30 seconds when a source is enabled
- Verified that an AccessControlException isn't thrown when doing `subscription:list` with an untrusted subscription URL.
